File: Backend/DTOs/ComponentSimulationResultDTO.cs
using System.Text.Json.Serialization;

namespace CircuitBench.DTOs
{
    [JsonPolymorphic(TypeDiscriminatorPropertyName = "kind")]
    [JsonDerivedType(typeof(ResistorSimulationResultDTO), "resistor")]
    [JsonDerivedType(typeof(LEDSimulationResultDTO), "led")]
    public abstract record ComponentSimulationResultDTO(double? Current, double? VoltageDrop, double? Power);

    public record ResistorSimulationResultDTO(double Resistance, double? Current, double? VoltageDrop, double? Power)
        : ComponentSimulationResultDTO(Current, VoltageDrop, Power);

    public record LEDSimulationResultDTO(double? Current, double? VoltageDrop, double? Power)
        : ComponentSimulationResultDTO(Current, VoltageDrop, Power);
}

File: Frontend/src/dtos/component-simulation-result-dto.ts
export interface BaseComponentSimulationResultDTO {
  readonly current: number | null
  readonly voltageDrop: number | null
  readonly power: number | null
}

export interface LEDSimulationResultDTO extends BaseComponentSimulationResultDTO {
  readonly kind: "led"
}

export interface ResistorSimulationResultDTO extends BaseComponentSimulationResultDTO {
  readonly kind: "resistor"
  readonly resistance: number
}

export type ComponentSimulationResultDTO =
  LEDSimulationResultDTO | ResistorSimulationResultDTO
